File: array/191.cpp
#include <bits/stdc++.h>

using namespace std;

void sortKSortedArray(vector<int> &arr, int k) {
    priority_queue<int, vector<int>, greater<int>> pq;

    for (int i = 0; i <= k; ++i) {
        pq.push(arr[i]);
    }

    int index = 0;
    for (int i = k + 1; i < arr.size(); ++i) {
        arr[index++] = pq.top();
        pq.pop();

        pq.push(arr[i]);
    }

    while(!pq.empty()) {
        arr[index++] = pq.top();
        pq.pop();
    }
}

int main() {
    vector<int> arr = {3, 4, 1, 2, 6, 7, 5, 10, 8, 9};

    sortKSortedArray(arr, 2);

    for (const auto &elem : arr) cout << elem << " ";

    return 0;
}

File: array/191.py
from typing import List
import heapq


def sort_k_sorted_array(arr: List, k: int):
    heap = []

    for index in range(0, k + 1):
        heapq.heappush(heap, arr[index])

    i = 0
    for index in range(k + 1, len(arr)):
        arr[i] = heapq.heappop(heap)
        i += 1

        heapq.heappush(heap, arr[index])

    while len(heap) > 0:
        arr[i] = heapq.heappop(heap)
        i += 1

    return arr


arr = [3, 4, 1, 2, 6, 7, 5, 10, 8, 9]
print(sort_k_sorted_array(arr, 2))
